### DISCUSSION TITLE
Are any "annotations" required on columns or is it all handled by crsql?

### DISCUSSION THREAD

**Discussion by jeromegn** [vlcn-io/cr-sqlite#187] [Q&A] [answered]
Here I'm wondering if the `LWW` example means I should set at least 1 column as a `LWW` to determine when to update a row?

Or is this handled by crsql? I assume this means changes should only be applied in order.

I'm building the next version of a project that will send "changes" that can be received in any order by peers. Does that mean I can't apply updates unless they're exactly +1 of the last known version from a peer?

**Comment by tantaman** [comment, marked as answer]
All the merge logic is handled for you. The only thing you need to ensure
is that when you record having seen changes up to a given database version
from a peer, you request the next set of changes from that peer starting at
(or before) that same version.

Let me know if this helps:
https://vlcn.io/networking/whole-crr-sync

**Reply by tantaman** [reply]
You can also check out the p2p reference implementation. It is pretty simple and short:

https://github.com/vlcn-io/cr-sqlite/tree/main/js/packages/p2p

in use by this example:

https://github.com/vlcn-io/live-examples/tree/main/src/todomvc-p2p

Much simpler than the client-server reference as the p2p protocol uses a "poke" scheme where peers let other peers know when data is available rather than streaming the data.
